src/sinks/util/buffer/metrics.rs

```diff
@@ -0,0 +1,523 @@
+use crate::event::{Event, Metric};
+use crate::sinks::util::Batch;
+use indexmap::IndexMap;
+
+trait Aggregate {
+    fn partition(&self) -> String;
+}
+
+impl Aggregate for Metric {
+    fn partition(&self) -> String {
+        match self {
+            Metric::Counter { name, tags, .. } => format!("{}{:?}", name, tags),
+            Metric::Gauge { name, tags, .. } => format!("{}{:?}", name, tags),
```

Comment:
We probably need to take the type into account here. I'm also not sure that generating strings like this is the best approach.

Reply:
> I'm also not sure that generating strings like this is the best approach.

I _feel_ like I understand what you mean, but maybe you could elaborate a bit? I've tried few times, but all my attempts produced much more verbose code then the 'string-typed' one.

Reply:
Yes, that was not the most helpful advice, sorry :smile: 

I've been thinking about this a quite a bit and I _think_ I have a promising solution:

```rust
struct MetricEntry(Metric);

impl Hash for MetricEntry {
    fn hash<H: Hasher>(&self, state: &mut H) {
        mem::discriminant(&self.0).hash(state);

        match &self.0 {
            Metric::Counter { name, tags, .. } => {
                name.hash(state);
                tags.hash(state);
            }
            Metric::Gauge { name, tags, .. } => {
                name.hash(state);
                tags.hash(state);
            }
            ...
        }
    }
}

struct MetricBuffer {
    metrics: IndexSet<MetricEntry>,
}
```

So we basically use a newtype to control how metrics are hashed, which lets us use all the existing hashmap-style functionality to our advantage.

Then in the `push` implementation we can do roughly this:

```rust
if let (existing) = self.metrics.swap_take(new_metric) {
    let merged = existing.merge(new_metric);
    self.metrics.insert(merged)
} else {
    self.metrics.insert(new_metric);
}
```

What do you think?

Reply:
@lukesteensen I think this will work, thanks! Had to implement some traits to make it work for floats and hashmaps. Could you take a look at the latest commit?

Also this `swap_take` operation shuffles metrics order a bit, not sure if it's important for us though.